R/pmcmc.R

```diff
@@ -0,0 +1,1029 @@
+#'        data time point.
+#'      - Secondly, to produce an unbiased estimate of the likelihood of the data given the proposed parameters.
+#' The log posterior of the initial parameters given the data is then estimated by adding the log prior and
+#' log likelihood estimate.
+#'
+#' The pMCMC sampler then proceeds as follows, for n_mcmc iterations:
+#' At each loop iteration the pMCMC sampler pefsorms three steps:
+#'   1. Propose new candidate samples for beta_start, beta_end and start_date based on
+#'     the current samples, using the proposal distribution
+#'     (currently multivariate Gaussian with user-input covariance matrix (proposal_kernel), and reflecting boundaries defined by pars_min, pars_max)
+#'   2. Calculate the log prior of the proposed parameters,
+#'      Use the particle filter to estimate log likelihood of the data given the proposed parameters, as described above,
+#'      as well as proposing a model state space.
+#'      Add the log prior and log likelihood estimate to estimate the log posterior of the proposed parameters given the data.
+#'   3. Metropolis-Hastings step: The joint canditate sample (consisting of the proposed parameters
+#'      and state space) is then accepted with probability min(1, a), where the acceptance ratio is
+#'      simply the ratio of the posterior likelihood of the proposed parameters to the posterior likelihood
+#'      of the current parameters. Note that by choosing symmetric proposal distributions by including
+#'      reflecting boundaries, we avoid the the need to include the proposal likelihood in the MH ratio.
+#'
+#'   If the proposed parameters and states are accepted then we update the current parameters and states
+#'   to match the proposal, otherwise the previous parameters/states are retained for the next iteration.
+#'
+#' After generating the pMCMC simulation, there are \code{replicates} specific iterations sampled based on the
+#' posterior probability. The parameters from those iterations are then used to generate new trajectories within
+#' the \code{squire_model} framework.
+#'
+#' @export
+#' @import coda
+#' @importFrom stats rnorm
+#' @importFrom mvtnorm rmvnorm
+
+pmcmc <- function(data,
+                  n_mcmc,
+                  log_likelihood = NULL,
+                  log_prior = NULL,
+                  n_particles = 1e2,
+                  steps_per_day = 4,
+                  output_proposals = FALSE,
+                  n_chains = 1,
+                  squire_model = explicit_model(),
+                  pars_obs = list(phi_cases = 0.1,
+                                  k_cases = 2,
+                                  phi_death = 1,
+                                  k_death = 2,
+                                  exp_noise = 1e6),
+                  pars_init = list('start_date'     = as.Date("2020-02-07"),
+                                   'R0'             = 2.5,
+                                   'Meff_dl'        = 2,
```

Comment:
You will hate me but could we make it Meff and Meff_pl. If only because Meff_dl is also before and during lockdown

Reply:
check